src/states/elder/mod.rs

```diff
@@ -1291,32 +1316,39 @@ impl Elder {
         }
     }
 
-    fn check_signed_message_trust(&self, msg: &SignedRoutingMessage) -> Result<(), RoutingError> {
-        if msg.check_trust(&self.chain) {
-            Ok(())
-        } else {
-            log_or_panic!(
-                LogLevel::Error,
-                "{} - Untrusted {:?} --- [{:?}]",
-                self,
-                msg,
-                self.chain.get_their_keys_info().format(", ")
-            );
-            Err(RoutingError::UntrustedMessage)
-        }
-    }
-
     // Signs and proves the given `RoutingMessage` and wraps it in `SignedRoutingMessage`.
     fn to_signed_routing_message(
         &self,
         msg: RoutingMessage,
+        node_knowledge_override: Option<u64>,
     ) -> Result<SignedRoutingMessage, RoutingError> {
-        let proof = self.chain.prove(&msg.dst, None);
+        let proof = self.chain.prove(&msg.dst, node_knowledge_override);
         let pk_set = self.our_section_bls_keys().clone();
         let secret_key = self.chain.our_section_bls_secret_key_share()?;
 
         SignedRoutingMessage::new(msg, secret_key, pk_set, proof)
     }
+
+    // Starts tracking knowledge of recently demoted elders.
+    fn update_demoted_members_knowledge(&mut self, removed_elders: &BTreeSet<P2pNode>) {
```

Comment:
I do not think this is correct, the member may not have processed it.
I think the correct way would be to wait for a message that the demoted elder send, so the demoted elder should send that message too.

Reply:
> the member may not have processed it yet.
Should be fine, because that means the member is still in the Elder state so when we send it `GenesisUpdate`, they would ignore it. Although maybe they should instead backlog it (and process when adult)?

Reply:
But then they would ignore a GenesisUpdate that they should not ignore which would be problematic (Also could be other messages like relocation).

What I am saying is I think this function break the logic behind the member_knowledge map (That is value in the map will always be trusted by the node).

Reply:
> But then they would ignore a GenesisUpdate that they should not ignore

Yes, which is why I'm thinking we should backlog it.

Reply:
But I do take you other point. But it won't solve the problem of lagging elder receiving `GenesisUpdate`.

Reply:
Yes, but we would only need to backlog message that we trusted as oppose to backlog any given random messages. (i.e message we really need to take action on).

Reply:
OK. So if we don't do this and only update the `members_knowledge` container when we receive the `MemberKnowledge` message, and also if we backlog `GenesisUpdate`, but only after successful proof check, then I think it should work. The worst case I can think of right now is this:

1. A node is demoted
2. It's lagging behind, so we don't receive it's `MemberKnowledge` yet. That means there is no entry for it in the `members_knowledge` container.
3. We progress the section again and send `GenesisUpdate` to all non-elders
4. Because there is no entry for the lagging node, we use the longest proof we have
5. The lagging node still thinking it is elder receives the message. It should trust it, because the proof is sufficiently long
6. It backlogs it
7. It eventually catches up, transitions to adult and processes the backlogged `GenesisUpdate`

So the only issue is the proof might get unnecessarily long, but that is probably OK.

Reply:
Alternative: on demote, instead of inserting the members_knowledge of the current version, we use the last version that reached the total consensus via the `AckMessage` event.

Reply:
Yes, both of these could be option, I'm a bit tired so I'll probably need to conclude tomorrow for definitive answer, but here is my thought.

I'm not clear why we need to remove the value from member_knowledge, it seem they should be kept up to date by sending a new message like all Adults do everytime as an elder we see Parsec change (i.e when we send GenesisUpdate). It just happen that when dealing with an elder we do not want to use that value (so we could check if it is an elder and send latest parsec), except in the case of Poke in which case the node explicitely tells us it still needs the old value. So I'm thinking the special poke case need to still exists, and in the non poke case, we really send latest parsec always (i.e the code is more like it was before).

Then regarding GenesisUpdate messages proofs (or other messages), we always use the member_knowledge, beause we always have it, and like for SMD it means every message we sent will always have enough proof to be trusted at the time we sent them.

But I'll need to think a bit more about it.

Reply:
>  So I'm thinking the special poke case need to still exists, and in the non poke case, we really send latest parsec always (i.e the code is more like it was before).

Or maybe not, and we use the value that was last given to us, i.e if we have not received the updated info, then the node is probably lagging and we send the old parsec in our gossip, but if we see it we send the new parsec, that should work for existing elders (lagging) as well as new elders (lagging or not).

Reply:
> I'm not clear why we need to remove the value from member_knowledge

We don't need that value once the node becomes elder, because for elder we prove messages based on the `their_knowledge` container. So the current logic I have there is to remove the value when the node is promoted and put it back when it is demoted. Also in the promotion case I don't remove it straight away because the node might be lagging. So I only remove it when I receive first up-to-date parsec gossip from the node.

Reply:
> We don't need that value once the node becomes elder

I think we are going at it from a different angle. I'm saying it seems simpler to continue to maintain that collection without differentiating between Elders and Adults. My reason is that it avoid the corner case that occur in transition from/to Elder. If at some place we do not need that value, we can choose to ignore it, even if we store it. (It is 7 nodes out of 100 so not much lost here).


I'm also thinking there are also different cases, i.e:
- When sending a section message to ourselves we use their-knowledge
- But when sending a message to one node (Like for example if we were sending a Relocation message), it would seem to make sense to use the members_knowledge (eventually), and the same machanism as GenesisUpdate message.


Additionally for:
```
     fn before_handle_parsec_gossip(&mut self, parsec_version: u64, sender: &PublicId) {
        if parsec_version >= self.parsec_map.last_version() { 
```

I am not sure that is even sufficient to provide the guarantee that we seem to assume here. i.e when participating in the DKG, we will send gossip, but we are not yet an elder, which mean we will not be in the group required to have voted for a change in their_knowledge.

Reply:
OK, I think I get your point now. It makes sense and looks like it makes things simpler, so :+1: